### DISCUSSION TITLE
semver is choosing pre-release version which is lower than the constraint

### DISCUSSION THREAD

**Discussion by PhilipAbed** [renovatebot/renovate#22539] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

using nuget 
i have this dependency: `Azure.Storage.Blobs : 12.3.0`
and i set the allowedVersions to `>= 12.13.0`

i debugged and found that after filtering versions per allowedVersions
```
[
  {
    "version": "12.13.0-beta.1",
    "releaseTimestamp": "2022-06-15T18:09:24.890Z"
  },
  {
    "version": "12.13.0",
    "releaseTimestamp": "2022-07-08T01:14:37.847Z"
  },
  ...
```

i get 12.13.0-beta.1 which is lower than the allowed version, 
after deeper investigation 
the problem is found here 
file:`filter.ts`
```
 filteredVersions = filteredVersions.filter((v) =>
        semver.satisfies(semver.coerce(v.version)!, allowedVersions)
      );
```

renovate has this in the middle `semver.coerce(v.version)!`
which is changing `12.13.0-beta.1` to `12.13.0` which makes the constraint meaningless.

after that renovate is slicing it to get the first available value, which satisfies the constraints.
file:`lookup/index.ts`
``` 
 if (isVulnerabilityAlert && !config.osvVulnerabilityAlerts) {
        filteredReleases = filteredReleases.slice(0, 1);
      }
```

Summary:
actual: getting an update to `12.13.0-beta.1`
expected: update to version `12.13.0`

Reproduction:
https://github.com/StinkyLord/RenovateTutorial1/pull/13

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: migrated config (repository=StinkyLord/RenovateTutorial1)
       "config": {
         "branchPrefix": "whitesource-remediate/",
         "enabledManagers": ["nuget"],
         "forkProcessing": "enabled",
         "onboarding": false,
         "requireConfig": "optional",
         "vulnerabilityAlerts": {
           "enabled": false,
           "ignoreUnstable": false,
           "groupName": null,
           "schedule": [],
           "dependencyDashboardApproval": false,
           "labels": ["security fix"],
           "commitMessageSuffix": ""
         },
         "extends": ["github>whitesource/merge-confidence:beta"],
         "packageRules": [
           {"matchPackagePatterns": ["*"], "enabled": false},
           {
             "matchDatasources": ["nuget"],
             "matchPackageNames": ["Azure.Storage.Blobs", "azure.storage.blobs"],
             "matchCurrentVersion": "12.3.0",
             "allowedVersions": ">= 12.13.0",
             "vulnerabilityAlert": true,
             "enabled": true,
             "ignoreUnstable": false,
             "groupName": null,
             "schedule": [],
             "dependencyDashboardApproval": false,
             "labels": ["security fix"],
             "prBodyTemplate": "{{{table}}}{{{notes}}}{{{changelogs}}}{{{controls}}}",
             "prBodyNotes": [
               "By merging this PR, the below vulnerabilities will be automatically resolved:\n\n| Severity           | 
<a href='#'><img src='https://whitesource-resources.whitesourcesoftware.com/cvss3.png' width='19' height='20'></a> CVSS 
Score       |CVE  |\n|---|---|---|\n| ![Medium](https://whitesource-resources.whitesourcesoftware.com/medium_vul_bb.png?
) Medium | 4.7 | [CVE-2022-30187](https://www.mend.io/vulnerability-database/CVE-2022-30187) |\n"
             ],
             "isVulnerabilityAlert": true,
             "commitMessageSuffix": "",
             "suppressNotifications": ["prIgnoreNotification"]
           }
         ]
       }
...
DEBUG: packageFiles with updates (repository=StinkyLord/RenovateTutorial1, baseBranch=main)
       "config": {
         "nuget": [
           {
             "deps": [
               {
                 "datasource": "nuget",
                 "depType": "nuget",
                 "depName": "Azure.Storage.Blobs",
                 "currentValue": "12.3.0",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "12.13.0-beta.1",
                     "newValue": "12.13.0-beta.1",
                     "releaseTimestamp": "2022-06-15T18:09:24.890Z",
                     "newMajor": 12,
                     "newMinor": 13,
                     "updateType": "minor",
                     "branchName": "whitesource-remediate/f398fc"
                   }
                 ],
                 "packageName": "Azure.Storage.Blobs",
                 "warnings": [],
                 "versioning": "nuget",
                 "sourceUrl": "https://github.com/Azure/azure-sdk-for-net",
                 "registryUrl": "https://api.nuget.org/v3/index.json",
                 "homepage": "https://github.com/Azure/azure-sdk-for-net/blob/Azure.Storage.Blobs_12.16.0/sdk/storage/Az
ure.Storage.Blobs/README.md",
                 "currentVersion": "12.3.0",
                 "isSingleVersion": true,
                 "fixedVersion": "12.3.0"
               },   
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
This was added long ago: https://github.com/renovatebot/renovate/commit/396fb977c35fe828f069f153ae8ea621767a796c

I think we can remove it, but we might need to do a validity check first: `semver.valid(v.version) && semver.satisfies(v.version)!, allowedVersions)`

**Reply by viceice** [reply]
wonder why semver does that strange conversion, as the sample is a valid semver v2

**Reply by msw-kialo** [reply]
@rarkins @viceice 

What is the recommendation if I "relied" on this behavior to filter docker tags (where all versions appear as prerelease and are now ignored)?

Use case: I set allowedVersion="<3.12" to define when Python docker images are updated on the next main version (e.g. `3.11.4-slim` is always updated to `3.11.5-slim` but only `3.12.0-slim` when we change the renovate config).

Semver has special handling for [prerelease tags](https://www.npmjs.com/package/semver#user-content-prerelease-tags): they only satisfy a range if the full version matches. `3.11.5-slim` does not satisfy `< 3.12.0` or even `< 3.12.0-slim`. While it might make sense in the context of NPM, this fallback is also used in other contexts and is confusing (e.g. `3.11-slim` and `3.11.5` still satisfy `< 3.12`).

One option would IMO invoking `semver.satisfies` with `includePrerelease: true`:

> Note that this behavior can be suppressed (treating all prerelease versions as if they were normal versions, for the purpose of range matching) by setting the includePrerelease flag on the options object to any [functions](https://github.com/npm/node-semver#functions) that do range matching.

like:

```ts
	 semver.satisfies(
          semver.valid(v.version) ? v.version : semver.coerce(v.version)!,
          allowedVersions,
    	  {includePrerelease: true}
     )
```

It would still ensure `12.13.0-beta.1` is not satisfied by `>= 12.13.0`, while supporting non-npm usages like `3.11.5-slim` satisfies `< 3.12`.

**Reply by rarkins** [reply]
Use a regex `allowedVersions` something like `/^3\.11\./`. Docker versioning will take care of enforcing the `-slim` suffix.

If this doesn't work, please create a reproduction repo to demonstrate

**Comment by viceice** [comment, marked as answer]
@PhilipAbed Can you please test this:

- https://github.com/renovatebot/renovate/pull/22586

**Reply by PhilipAbed** [reply]
i have tested 
same repo, run after the other: see 
old : https://github.com/StinkyLord/RenovateTutorial1/pull/13
new : https://github.com/StinkyLord/RenovateTutorial1/pull/14
